### ISSUE TITLE
File removed in before_destroy is missing if destroy fails and rolls back

### ISSUE THREAD

**Issue by rlow** [markevans/dragonfly#477]
If a model destroy action fails and the database rolls back the transaction, the ```before_destroy``` callback has already removed the file, leaving the model data and file datastore in an inconsistent state.

Could this be fixed by changing the callback to ```after_destroy```?

**Comment by markevans** [comment]
potentially yes. what's an example of destroy failing? (so I can add a test)

**Comment by rlow** [comment]
We have a model with a `has_many` relation with a database foreign_key constraint (Postgres in our case). On our model `has_many` association we have `dependent: :restrict_with_error` which handles catching the dependent error apparently before the destroy happens. To cause this issue, you can remove the `dependent:` option altogether, and let the database constraint cause a failure and rollback. 

Thanks

**Comment by markcell** [comment]
Hi, I also came across this situation. Is there any forecast for an update that includes this change?

Thanks.

**Comment by markevans** [comment]
released in version 1.1.5 - cheers

**Comment by dlibanori** [comment]
Actually `after_destroy` is called *before* commit. As @rlow said, `before_destroy` is not safe and removing `dependent` will call `destroy_dragonfly_attachments` before `DELETE`. But this is not related with any transaction, SQL fails just because foreign key constraint's fails, not because a `COMMIT`. The real issue is when you start a transaction, calls `destroy` then rollback it. Example:

```
class Foo < ApplicationRecord
  dragonfly_accessor :bar
end

f = Foo.create(bar: File.new('some.blob'))
Foo.transation do
  f.destroy
  raise ActiveRecord::Rollback
end

# Ooops! There is no more file :'(
```
